### PR TITLE
[fitting] add mean normal derivative

### PR DESCRIPTION
**Pull request by Gowthraven**

### CODE REVIEW

**Comment by nmellado** [comment]
Thanks for the proposal.
Could you please add some documentation explaining what this class does, as for `MeanPositionDer` ?

File: Ponca/src/Fitting/mean.h [hunk 1]
```diff
@@ -95,33 +95,77 @@ namespace Ponca {
...
-        /// We have \f$ s'(\mathbf{x}) = \sum_i w'_\mathbf{x}(\mathbf{p_i}) \f$.
-        ///
-        /// We rewrite \f$ t(\mathbf{x}) = \sum u(\mathbf{x})v(\mathbf{x}) \f$, with \f$ u(\mathbf{x}) = w_\mathbf{x}(\mathbf{p_i}) \f$ and \f$ v(\mathbf{x}) = \mathbf{p_i} \f$.
-        ///
-        /// As the point cloud coordinates are constants, \f$v(\mathbf{x})\f$ is constant, its derivative is null, and so \f$  t'(\mathbf{x}) = \sum_i u'(\mathbf{x}) v(\mathbf{x}) = \sum_i w'_\mathbf{x}(\mathbf{p_i}) \mathbf{p_i} \f$.
-        ///
-        /// Which leads to \f$ b'(\mathbf{x}) = \frac{\sum_i w'_\mathbf{x}(\mathbf{p_i}) \mathbf{p_i} - b(\mathbf{x})\sum w'(\mathbf{x})}{\sum_i w_\mathbf{x}(\mathbf{p_i})} \f$
-        ///
-        /// \note This code is not directly tested, but rather indirectly by testing CovariancePlaneDer::dNormal()
+        // / \brief Compute derivatives of the barycenter.
```
Comment (nmellado) [inline Ponca/src/Fitting/mean.h:98]: Please stick to `///` for this line and the next ones
```suggestion
        /// \brief Compute derivatives of the barycenter.
```

**Review by nmellado — CHANGES_REQUESTED** [review]
Thx !
I need to compile the doc to check the result (just don't forget to keep `///` for comment lines).

File: Ponca/src/Fitting/mean.h (hunk 1 above)
Comment (Gowthraven) [reply Ponca/src/Fitting/mean.h]: I have updated the comments on my last commit.

File: Ponca/src/Fitting/mean.h [hunk 2]
```diff
@@ -122,6 +122,50 @@ namespace Ponca {
...
     }; //class MeanPositionDer
 
+    template<class DataPoint, class _WFunctor, int DiffType, typename T>
+    class MeanNormalDer : public T {
+        PONCA_FITTING_DECLARE_DEFAULT_TYPES
+        PONCA_FITTING_DECLARE_DEFAULT_DER_TYPES
+
+    protected:
+        enum {
+            Check = Base::PROVIDES_PRIMITIVE_DERIVATIVE and
```
Comment (nmellado) [inline Ponca/src/Fitting/mean.h:132]:
```suggestion
            Check = Base::PROVIDES_PRIMITIVE_DERIVATIVE && 
```

File: Ponca/src/Fitting/mean.h [hunk 3]
```diff
@@ -122,6 +122,50 @@ namespace Ponca {
...
+        PONCA_FITTING_DECLARE_DEFAULT_DER_TYPES
+
+    protected:
+        enum {
+            Check = Base::PROVIDES_PRIMITIVE_DERIVATIVE and 
+                    Base::PROVIDES_MEAN_NORMAL,
+            PROVIDES_MEAN_NORMAL_DERIVATIVE,  /*!< \brief Provides derivative of the mean normal*/
+        };
+
+        /*! \brief Derivatives of the input normals of the input points vectors*/
```
Comment (nmellado) [inline Ponca/src/Fitting/mean.h:137]:
```suggestion
    private:
        /*! \brief Derivatives of the input normals of the input points vectors*/
```

File: Ponca/src/Fitting/mean.h [hunk 4]
```diff
@@ -122,6 +122,50 @@ namespace Ponca {
...
+    /// Assuming the normal vectors themselves do not change with position, \f$v(\mathbf{x})\f$ is constant, its derivative is null, 
+    /// and so \f$ t'(\mathbf{x}) = \sum_i u'(\mathbf{x}) v(\mathbf{x}) = \sum_i w'_\mathbf{x}(\mathbf{n_i}) \mathbf{n_i} \f$.
+    ///
+    /// Which leads to \f$ n'(\mathbf{x}) = \frac{\sum_i w'\mathbf{x}(\mathbf{n_i}) \mathbf{n_i} - n(\mathbf{x})\sum w'(\mathbf{x})}{\sum_i w\mathbf{x}(\mathbf{n_i})} \f$.
+    
+    /// \note This code is not directly tested. 
+
+        PONCA_EXPLICIT_CAST_OPERATORS_DER(MeanNormalDer,meanNormalDer)
+        PONCA_FITTING_DECLARE_INIT
+        PONCA_FITTING_DECLARE_ADDNEIGHBOR_DER
```
Comment (nmellado) [inline Ponca/src/Fitting/mean.h:166]:
```suggestion
        PONCA_EXPLICIT_CAST_OPERATORS_DER(MeanNormalDer,meanNormalDer)
        PONCA_FITTING_DECLARE_INIT
        PONCA_FITTING_DECLARE_ADDNEIGHBOR_DER

        /// \brief Compute the derivative of the mean normal vector of the input points. 
        /// 
        /// ### Step-by-step derivation of the mean normal : 
        ///  Given the definition of the mean normal \f$ n(\mathbf{x}) = \frac{\sum_i w_\mathbf{x}(\mathbf{n_i}) \mathbf{n_i}}{\sum_i w_\mathbf{x}(\mathbf{n_i})} \f$, 
        /// where \f$\left[\mathbf{n_i} \in \text{neighborhood}(\mathbf{x})\right]\f$ are all the normal samples in \f$\mathbf{x}\f$'s neighborhood.
        ///
        /// We denote \f$ t(\mathbf{x}) = \sum_i w_\mathbf{x}(\mathbf{n_i}) \mathbf{n_i} \f$ and \f$ s(\mathbf{x}) = \sum_i w_\mathbf{x}(\mathbf{n_i})\f$, 
        /// such that \f$ n(\mathbf{x}) = \frac{t(\mathbf{x})}{s(\mathbf{x})}\f$.
        ///
        /// By definition, \f$ n'(\mathbf{x}) = \frac{s(\mathbf{x})t'(\mathbf{x}) - t(\mathbf{x})s'(\mathbf{x})}{s(\mathbf{x})^2}\f$.
        ///
        /// Assuming the weight of each normal is dependent on the position, we have \f$ s'(\mathbf{x}) = \sum_i w'_\mathbf{x}(\mathbf{n_i}) \f$.
        ///
        /// We rewrite \f$ t(\mathbf{x}) = \sum u(\mathbf{x})v(\mathbf{x}) \f$, with \f$ u(\mathbf{x}) = w_\mathbf{x}(\mathbf{n_i}) \f$ and \f$ v(\mathbf{x}) = \mathbf{n_i} \f$.
        ///
        /// Assuming the normal vectors themselves do not change with position, \f$v(\mathbf{x})\f$ is constant, its derivative is null, 
        /// and so \f$ t'(\mathbf{x}) = \sum_i u'(\mathbf{x}) v(\mathbf{x}) = \sum_i w'_\mathbf{x}(\mathbf{n_i}) \mathbf{n_i} \f$.
        ///
        /// Which leads to \f$ n'(\mathbf{x}) = \frac{\sum_i w'\mathbf{x}(\mathbf{n_i}) \mathbf{n_i} - n(\mathbf{x})\sum w'(\mathbf{x})}{\sum_i w\mathbf{x}(\mathbf{n_i})} \f$.    
        /// \note This code is not directly tested. 
        PONCA_MULTIARCH inline const VectorArray& normalDerivatives() const { return m_dSumN; }
```

**Review by nmellado — COMMENTED** [review]
Small suggestions, and we are ready to go !

**Review by nmellado — APPROVED** [review]

**Pull request merged**